controllers/atlasmigration_controller.go

```diff
@@ -16,39 +16,310 @@ limitations under the License.
+	if err := r.Get(ctx, client.ObjectKey{Namespace: ns, Name: selector.Name}, secret); err != nil {
+		return "", err
+	}
+
+	us := string(secret.Data[selector.Key])
+	return us, nil
+}
+
+// createTmpDir creates a temporary directory and returns its url.
+func (r *AtlasMigrationReconciler) createTmpDir(
+	ctx context.Context,
+	ns string,
+	dir dbv1alpha1.Dir) (string, func() error, error) {
+	// Get configmap
+	configMap := corev1.ConfigMap{}
+	if err := r.Get(ctx, types.NamespacedName{
+		Namespace: ns,
+		Name:      dir.ConfigMapRef,
+	}, &configMap); err != nil {
+		return "", nil, err
+	}
+
+	// Create temporary directory and remove it at the end of the function
+	tmpDir, err := ioutil.TempDir("", "migrations")
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Foreach configmap to build temporary directory
+	// key is the name of the file and value is the content of the file
+	for key, value := range configMap.Data {
+		filePath := filepath.Join(tmpDir, key)
+		err := ioutil.WriteFile(filePath, []byte(value), 0644)
+		if err != nil {
+			// Remove the temporary directory if there is an error
+			os.RemoveAll(tmpDir)
+			return "", nil, err
+		}
+	}
+
+	return fmt.Sprintf("file://%s", tmpDir), func() error {
+		return os.RemoveAll(tmpDir)
+	}, nil
+}
+
+// Update the status of the given AtlasMigration resource.
+// If err is not nil, the status will be set to false and the reason will be set to the error message.
+func (r *AtlasMigrationReconciler) updateResourceStatus(
+	ctx context.Context,
+	am dbv1alpha1.AtlasMigration, err error) error {
+	conditionStatus := metav1.ConditionTrue
+	conditionMessage := ""
+	// Be careful when editing this default value, it required when updating the status of the resource
+	conditionReason := "Applied"
+
+	if err != nil {
+		conditionStatus = metav1.ConditionFalse
+		conditionReason = "Reconciling"
+		conditionMessage = strings.TrimSpace(err.Error())
+	}
+
+	meta.SetStatusCondition(
+		&am.Status.Conditions,
+		metav1.Condition{
+			Type:    "Ready",
+			Status:  conditionStatus,
+			Reason:  conditionReason,
+			Message: conditionMessage,
+		},
+	)
+
+	if err := r.Status().Update(ctx, &am); err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *AtlasMigrationReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&dbv1alpha1.AtlasMigration{}).
 		Complete(r)
 }
+
+// executeHCLTemplate executes the given HCL template and returns the path of the temporary file.
+func executeHCLTemplate(tmplName string, data interface{}) (string, func() error, error) {
+	var buf bytes.Buffer
+	if err := tmpl.ExecuteTemplate(&buf, tmplName, data); err != nil {
+		return "", nil, err
+	}
+
+	return atlas.TempFile(buf.String(), "hcl")
+}
+
+// Pernament errors are defined as:
+// - K8s Client NotFound error
+// - SQL error
+// - Invalid Yaml error
+func isPernamentError(err error) bool {
```

Comment:
IMO, better to make this decision explicitly near where the call is made. it may cause some duplication, but the code will be much easier to follow. in addition, there may be cases where the same err can have different meanings in different contexts

Reply:
and read this when you have time:
https://rotemtam.com/2020/05/18/the-dry-principle-is-bad-advice/